### DISCUSSION TITLE
TypeScript Snapshot type

### DISCUSSION THREAD

**Discussion by andraaspar** [pmndrs/valtio#844] [Q&A] [answered]
The [documentation for useSnapshot](https://valtio.pmnd.rs/docs/api/basic/useSnapshot) provides the following JavaScript example:
```js
function AuthorView() {
  const snap = useSnapshot(state)
  return (
    <div>
      {snap.books.map((book, i) => (
        <Book key={book.id} bookProxy={state.books[i]} bookSnapshot={book} />
      ))}
    </div>
  )
}
```
What would be the correct type for the `Book` component's `bookSnapshot` parameter in TypeScript? I guess `INTERNAL_Snapshot<T>` should not be used, because it is internal.

**Comment by dai-shi** [comment, marked as answer]
I would simply use the same type as non-snapshot book.
`INTERNAL_Snapshot<T>` is trying to be super precise, but it's usually not necessary, and because you know the exact type of book, it's fair.

It became somewhat stable and I may reconsider exporting `Snapshot<T>` type in v2. 🤔

**Reply by andraaspar** [reply]
Thank you for the answer.

+1 for exporting the `Snapshot<T>` in v2, as I would not need to cast my `Snapshot<IBook>` to `IBook` and I could still have the nice read only warnings inside the `Book` component.
